connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectSessionManager.scala

```diff
@@ -95,47 +92,134 @@ class SparkConnectSessionManager extends Logging {
     Option(getSession(key, None))
   }
 
-  private def getSession(
-      key: SessionKey,
-      default: Option[Callable[SessionHolder]]): SessionHolder = {
-    val session = default match {
-      case Some(callable) => sessionStore.get(key, callable)
-      case None => sessionStore.getIfPresent(key)
-    }
-    // record access time before returning
-    session match {
-      case null =>
-        null
-      case s: SessionHolder =>
-        s.updateAccessTime()
-        s
+  private def getSession(key: SessionKey, default: Option[() => SessionHolder]): SessionHolder = {
+    sessionsLock.synchronized {
+      // try to get existing session from store
+      val sessionOpt = sessionStore.get(key)
+      // create using default if missing
+      val session = sessionOpt match {
+        case Some(s) => s
+        case None =>
+          default match {
+            case Some(callable) =>
+              val session = callable()
+              sessionStore.put(key, session)
+              session
+            case None =>
+              null
+          }
+      }
+      // record access time before returning
+      session match {
+        case null =>
+          null
+        case s: SessionHolder =>
+          s.updateAccessTime()
+          s
+      }
     }
   }
 
   def closeSession(key: SessionKey): Unit = {
-    // Invalidate will trigger RemoveSessionListener
-    sessionStore.invalidate(key)
-  }
-
-  private class RemoveSessionListener extends RemovalListener[SessionKey, SessionHolder] {
-    override def onRemoval(notification: RemovalNotification[SessionKey, SessionHolder]): Unit = {
-      val sessionHolder = notification.getValue
-      sessionsLock.synchronized {
+    var sessionHolder: Option[SessionHolder] = None
+    sessionsLock.synchronized {
+      sessionHolder = sessionStore.remove(key)
+      sessionHolder.foreach { s =>
         // First put into closedSessionsCache, so that it cannot get accidentally recreated by
         // getOrCreateIsolatedSession.
-        closedSessionsCache.put(sessionHolder.key, sessionHolder.getSessionHolderInfo)
+        closedSessionsCache.put(s.key, s.getSessionHolderInfo)
       }
-      // Rest of the cleanup outside sessionLock - the session cannot be accessed anymore by
-      // getOrCreateIsolatedSession.
-      sessionHolder.close()
+    }
+    // Rest of the cleanup outside sessionLock - the session cannot be accessed anymore by
+    // getOrCreateIsolatedSession.
+    sessionHolder.foreach { s =>
+      s.close()
+      // Update in closedSessionsCache: above it wasn't updated with closedTime etc. yet.
```

Comment:
Any concerns if `close()` above throws? 
Btw, when do we clean up `closedSessionsCache`?

Reply:
Generally nothing really bad would happen, as this would be an unexpected exception, and the session is already removed from the manager.
If this is called from by periodic maintenance, it would be caught, and then next periodic maintenance will go on.
If this is called from ReleaseSession, an error will be thrown to the user, which I think is correct if something unexpected happened. We might potentially not clean something up fully, but this is in general an unexpected situation.

Reply:
Should we remove closedSessionsCache.put() below?

Reply:
We deliberetely put it twice
- the first time, under lock prevents the session to be recreated, but the session is not fully closed at that point
- the second time, after it's finished shutting down, updates the record with a final record that has the final state of the session (closed time etc.)